src/System.IO.Pipelines/PipeCompletion.cs

```diff
@@ -1,58 +1,137 @@
 
-        private Exception _exception;
 #if COMPLETION_LOCATION_TRACKING
         private string _completionLocation;
 #endif
+        public Exception Exception { get; set; }
 
-        public bool IsCompleted => _exception != null;
+        private PipeCompletionCallback[] _callbacks;
+        private int _callbackCount;
 
-        public void TryComplete(Exception exception = null)
+        public string Location
+        {
+            get
+            {
+#if COMPLETION_LOCATION_TRACKING
+                return _completionLocation;
+#else
+                return null;
+#endif
+            }
+        }
+
+        public bool IsCompleted => Exception != null;
+
+        public bool TryComplete(Exception exception = null)
         {
 #if COMPLETION_LOCATION_TRACKING
             _completionLocation = Environment.StackTrace;
 #endif
-            // Set the exception object to the exception passed in or a sentinel value
-            Interlocked.CompareExchange(ref _exception, exception ?? _completedNoException, null);
+            if (Exception == null)
+            {
+                // Set the exception object to the exception passed in or a sentinel value
+                Exception = exception ?? _completedNoException;
+            }
+            return _callbackCount > 0;
         }
 
-        public bool IsCompletedOrThrow()
+        public bool AddCallback(Action<Exception, object> callback, object state)
         {
-            if (_exception != null)
+            if (IsCompleted)
             {
-                if (_exception != _completedNoException)
-                {
-                    ThrowFailed();
-                }
                 return true;
             }
+
+            if (_callbacks == null)
+            {
+                _callbacks = new PipeCompletionCallback[InitialCallbacksSize];
+            }
+
+            var newIndex = _callbackCount;
+            _callbackCount++;
+
+            if (_callbackCount == _callbacks.Length)
+            {
+                var newArray = new PipeCompletionCallback[_callbacks.Length * 2];
+                Array.Copy(_callbacks, newArray, _callbacks.Length);
+                _callbacks = newArray;
+            }
+            _callbacks[newIndex].Callback = callback;
+            _callbacks[newIndex].State = state;
             return false;
         }
 
-        private void ThrowFailed()
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public bool IsCompletedOrThrow()
         {
-            throw _exception;
+            if (Exception == null)
+            {
+                return false;
+            }
+
+            if (Exception != _completedNoException)
+            {
+                ThrowFailed();
+            }
+
+            return true;
         }
 
-        public string Location
+        public void InvokeCallbacks()
         {
-            get
+            Debug.Assert(IsCompleted);
+
+            if (_callbacks == null)
```

Comment:
It doesn't look like this method is being called with a lock which seems very bad.